classes/task/sendfailednotifications_task.php

```diff
@@ -0,0 +1,168 @@
+
+use tool_log\log\manager;
+use logstore_xapi\log\store;
+
+class sendfailednotifications_task extends \core\task\scheduled_task {
+
+    /**
+     * Get a descriptive name for this task (shown to admins).
+     *
+     * @return string
+     */
+    public function get_name() {
+        return get_string('tasksendfailednotifications', 'logstore_xapi');
+    }
+
+    /**
+     * Get failed rows.
+     *
+     * @return array
+     */
+    private function get_failed_rows() {
+        global $DB;
+
+        $sql = "SELECT x.id, x.errortype AS type, x.eventname, u.firstname, u.lastname, x.contextid, x.response, x.timecreated
+        FROM {logstore_xapi_failed_log} x
+        LEFT JOIN {user} u
+        ON u.id = x.userid";
+
+        $results = $DB->get_records_sql($sql);
+        return $results;
+    }
+
+    /**
+     * Get failed email message.
+     *
+     * @return string email message in html
+     */
+    private function get_failed_email_message($results) {
+        $emailmsg = "";
+
+        $emailmsg .= "<p>".PHP_EOL;
+        $emailmsg .= get_string('failedtosend', 'logstore_xapi').PHP_EOL;
+        $emailmsg .= "</p>".PHP_EOL;
+
+        // summary info
+        $emailmsg .= "<table>".PHP_EOL;
+
+        $endpointname = get_string('endpoint', 'logstore_xapi');
+        $url = get_config('logstore_xapi', 'endpoint');
+        $endpointurl = "<a href='$url' target='_blank'>$url</a>";
+
+        $errorlogpage = get_string('errorlogpage', 'logstore_xapi');
+        $url = new \moodle_url("/admin/tool/log/store/xapi/report.php");
+        $errorlogurl = "<a href='$url' target='_blank'>$url</a>";
+
+        $row = "<tr>";
+        $row .= "<td>".$endpointname."</td>";
+        $row .= "<td>".$endpointurl."</td>";
+        $row .= "</tr>".PHP_EOL;
+        $emailmsg .= $row;
+
+        $row = "<tr>";
+        $row .= "<td>".$errorlogpage."</td>";
+        $row .= "<td>".$errorlogurl."</td>";
+        $row .= "</tr>".PHP_EOL;
+        $emailmsg .= $row;
+
+        $emailmsg .= "</table>".PHP_EOL;
+
+        // separator
+        $emailmsg .= "<h3>".get_string('failurelog', 'logstore_xapi')."</h3>".PHP_EOL;
+
+        // rows
+        $emailmsg .= "<table>".PHP_EOL;
+
+        $emailmsg .= '<tr>'.PHP_EOL;
+        $emailmsg .= "<th align='left'>".get_string('datetimegmt', 'logstore_xapi')."</th>".PHP_EOL;
+        $emailmsg .= "<th align='left'>".get_string('eventname', 'logstore_xapi')."</th>".PHP_EOL;
+        $emailmsg .= "<th align='left'>".get_string('response', 'logstore_xapi')."</th>".PHP_EOL;
+        $emailmsg .= "</tr>".PHP_EOL;
+
+        foreach ($results as $result) {
+            $row = "<tr>";
+            $row .= "<td>".userdate($result->timecreated)."</td>";
+            $row .= "<td>".$result->eventname."</td>";
+            $row .= "<td>".$result->response."</td>";
+            $row .= "</tr>".PHP_EOL;
+
+            $emailmsg .= $row;
+        }
+        $emailmsg .= "</table>".PHP_EOL;
```

Comment:
You should use `html_writer` for generating html
https://docs.moodle.org/dev/html_writer

Reply:
Agreed. 
Is it necessary to use PHP_EOL?

Reply:
> Agreed.
> Is it necessary to use PHP_EOL?

Yeah I was wondering about that too, but if it's changed to `html_writer` surely that takes care of newlines

Reply:
Perhaps  using html_table class should be better.

Reply:
I have changed the code to use html_writer and html_table avoiding the need for my own html formatting.